### PR TITLE
fix: Correctly handle errors in method streams.

### PR DESCRIPTION
**Pull request by SandPod**
Two issue where identified in the method stream interface:

### Catch non generator "Stream" exceptions
When a streaming endpoint returns a stream instead of being a generator and an exception occurs before the stream is returned our error handling didn't manage to catch the error.

This resulted in the stream not being closed on the client and the exception being propagated up on the server.

### No error logging
During a refactor for how streams are closed to not produce memory leaks, closing the session with any errors that occurred was accidentally removed. PR here: https://github.com/serverpod/serverpod/pull/2573
This PR reintroduces closing sessions with errors if one has occurred.

This is tested through the logging interface since that is the only output that displays if an error has occurred in a session.

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - these are stability and reporting fixes inside of the streaming interface.

### CODE REVIEW

File: tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart [hunk 1]
```diff
@@ -76,4 +76,38 @@ void main() {
...
+    stream.listen((event) {
+      // Do nothing
+    }, onError: (error) {
+      streamComplete.complete(error);
+    });
+
+    await expectLater(
+      await streamComplete.future,
+      isA<ConnectionClosedException>(),
+    );
```
Comment (hampuslavin) [inline tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart:94]: Question: Do you think it's possible to omit the completer and use built-in assertion for this? Same for next test.
```dart
await expectLater(stream, emitsError(isA<ConnectionClosedException>))
```

File: tests/serverpod_test_server/test_integration/logging/endpoint_method_stream_logging_test.dart [hunk 2]
```diff
@@ -142,6 +142,31 @@ void main() async {
...
+        isA<ConnectionClosedException>(),
+      );
+
+      // Wait for the log to be written
+      await Future.delayed(Duration(milliseconds: 100));
+      var logs = await LoggingUtil.findAllLogs(session);
+
+      expect(logs, hasLength(1));
+      var log = logs.firstOrNull;
+      expect(log?.sessionLogEntry.error, isNotNull);
```
Comment (hampuslavin) [inline tests/serverpod_test_server/test_integration/logging/endpoint_method_stream_logging_test.dart:168]: Nice!

**Review by hampuslavin — COMMENTED** [review]
Nice! 🐛 💣  Left a small suggestion for the test!

File: tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart (hunk 1 above)
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart]: Updated all the tests in the file to this pattern.

File: tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart [hunk 3]
```diff
@@ -76,4 +76,38 @@ void main() {
       isA<ExceptionWithData>(),
     );
   });
+
+  test(
+      'Given a streaming method that is not a generator that throws an exception before the stream is returned, then stream is closed with ServerpodClientException.',
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart]: Updated the description to have a when section, didn't format it like your suggestion but I think it should cover the feedback here.

File: tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart [hunk 4]
```diff
@@ -76,4 +76,38 @@ void main() {
...
+    });
+
+    await expectLater(
+      await streamComplete.future,
+      isA<ConnectionClosedException>(),
+    );
+  });
+
+  test(
+      'Given a streaming method that is not a generator that throws an exception as its first message when calling method, then stream is closed with ServerpodClientException.',
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/method_streaming/out_stream_test.dart:69]: See other comment.

**Review by christerswahn — APPROVED** [review]
LGTM!

**Review by hampuslavin — APPROVED** [review]
👏 👏

**Pull request merged**